File: tp_tda/tp2/aulas.py
from collections import deque

def bfs(start, tunnels):
    queue = deque([start]) 
    visited = {start: True}
    energy = [0]*len(tunnels)
    
    while queue:
        currentNode = queue.popleft()
        
        if currentNode+1 < len(tunnels) and currentNode+1 not in visited:   
            queue.append(currentNode+1)
            visited[currentNode+1] = True
            energy[currentNode+1] = energy[currentNode]+1
            
        if currentNode-1 >= 0 and currentNode-1 not in visited:   
            queue.append(currentNode-1)
            visited[currentNode-1] = True
            energy[currentNode-1] = energy[currentNode]+1
        
        if tunnels[currentNode]-1 != currentNode and tunnels[currentNode]-1 not in visited:
            queue.append(tunnels[currentNode]-1)
            energy[tunnels[currentNode]-1]= energy[currentNode]+1
            visited[tunnels[currentNode]-1] = True
                
    return energy

def main():
    n = int(input().strip())
    tunnels = list(map(int, input().strip().split()))
    
    res = bfs(0,tunnels)
        
    print(" ".join(map(str, res)))

main()

File: tp_tda/tp2/aulas.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <string>
#include <sstream>

using namespace std;

vector<long long> bfs(long long start, const vector<long long>& tunnels) {
    queue<long long> q;
    vector<bool> visited(tunnels.size(), false);
    vector<long long> energy(tunnels.size(), 0);
    
    q.push(start);
    visited[start] = true;
    
    while (!q.empty()) {
        long long currentNode = q.front();
        q.pop();
        
        if (currentNode + 1 < tunnels.size() && !visited[currentNode + 1]) {
            q.push(currentNode + 1);
            visited[currentNode + 1] = true;
            energy[currentNode + 1] = energy[currentNode] + 1;
        }
        
        long long tunnel_destination = tunnels[currentNode] - 1;
        if (tunnel_destination != currentNode && !visited[tunnel_destination]) {
            q.push(tunnel_destination);
            visited[tunnel_destination] = true;
            energy[tunnel_destination] = energy[currentNode] + 1;
        }
    }
    
    return energy;
}

int main() {
    long long n;
    cin >> n;
    
    vector<long long> tunnels(n);
    for (long long i = 0; i < n; ++i) {
        cin >> tunnels[i];
    }
    
    vector<long long> res = bfs(0, tunnels);

    for (long long i = 0; i < res.size(); ++i) {
        if (i > 0) cout << " ";
        cout << res[i];
    }
    cout << endl;
    
    return 0;
}
